### PR TITLE
No sharing targets

### PR DESCRIPTION
**Pull request by swalker2m**
This PR presents for discussion some ideas about how to change the API to remove target sharing.  Feedback is particularly encouraged here.

### CODE REVIEW

File: modules/core/src/main/scala/lucuma/odb/api/model/ValidatedInput.scala [hunk 1]
```diff
@@ -32,7 +32,7 @@ object ValidatedInput {
       case (None,     Some(bʹ)) => bʹ.asRight[A].some.validNec[InputError]
       case (Some(_),  Some(_) ) => InputError.fromMessage(s"Either $nameA or $nameB are permitted but not both").invalidNec
       case (None,     None    ) => Option.empty[Either[A, B]].validNec[InputError]
-    }.fold(_.invalid[Option[Either[A, B]]], identity)
```
Comment (swalker2m) [inline modules/core/src/main/scala/lucuma/odb/api/model/ValidatedInput.scala:35]: This is an unrelated change that sneaked in.

**Comment by toddburnside** [comment]
Since targets don't really have an existence independent of the observation, it seems to make sense to use a list of targets instead of Either[Target, Asterism]. 

I guess one result of this is that each top-level item in the targets tab tree list would be a list of one or more targets, and bulk editing would be for that list. This also allows dragging and dropping of observations to/from target lists. But, it would not allow any bulk editing for targets that are otherwise equal but reside in lists of targets that differ - this may be unavoidable.

**Review by cquiroz — APPROVED** [review]
Looks good, some semi smart comments included

File: docs/unshareTargets.md [hunk 2]
```diff
@@ -0,0 +1,396 @@
...
+`editSiderealScienceTarget` mutation could take a program ID, or a list of observation IDs, along 
+with a sidereal target name.  It would then apply provided updates directly.
+
+For example, set the radial velocity of all "NGC 5949" targets in program `p-34d5`.
+
+```json
+{
+  "editSiderealScienceTarget": {
+    "programId":      "p-34d5",
+    "name":           "NGC 5949",
```
[the remark this answers is not in the document]
Comment (toddburnside) [reply docs/unshareTargets.md:256]: We'd definitely have the same target name multiple times for different observations. And above, it notes that the name would have to be unique for all the targets within an observation. But, in order to prevent a problem with different observations having the targets with the same name but not otherwise being equal, I think this mutation might need to take a list of observation ids.

**Comment by swalker2m** [comment]
>  But, it would not allow any bulk editing for targets that are otherwise equal but reside in lists of targets that differ

That's an astute observation. It is true in Explore as currently conceived but I think we could always add mutations and grouping queries that work across differing lists in the API.  I tried to hint at it at least in the proposal with this input:

```json
 {
   "editSiderealScienceTarget": {
     "programId":      "p-34d5",
     "name":           "NGC 5949",
     "radialVelocity": { "metersPerSecond": 423607 }
   }
 }
 ```

the idea being that it would update all "NGC 5949" targets regardless of whether the other `Targets` information differs and even regardless of whether the "NGC 5949" instances themselves differ.  Similarly, we can discover groups of observations that have "NGC 5949" with queries for that purpose.

File: docs/unshareTargets.md (hunk 2 above)
[the remark this answers is not in the document]
Comment (toddburnside) [reply docs/unshareTargets.md:256]: Ah. You mentioned an optional list of observation ids above. That seems sufficient. It would be up to the user (api user or UI) to know if it made sense to do it for all targets of that name in the program.

**Review by toddburnside — APPROVED** [review]
I've tried to think of better alternatives, but failed. LGTM.

**Comment by toddburnside** [comment]
> It is true in Explore as currently conceived but I think we could always add mutations and grouping queries that work across differing lists in the API.

That's true. I guess we could start simple in the UI and evolve it as needed and time permits.

File: docs/unshareTargets.md [hunk 3]
```diff
@@ -0,0 +1,396 @@
...
+forward is not immediately obvious.  This document proposes a series of updates that accomplish 
+the goal while still supporting bulk editing.  
+
+## Target IDs
+
+Target IDs make otherwise identical targets distinct.  For example, two "NGC 5949" instances 
+would have distinct IDs, making them incompatible with grouping by target.  Instead of referring to 
+targets by ID, we'll instead use target names and require unique names within a target list.
+
+(Since the `name` field will be needed to identify a target to edit, this raises the question of
```
[the remark this answers is not in the document]
Comment (swalker2m) [reply docs/unshareTargets.md:13]: You raise some good points and I didn't put much thought into the name.  I'd just thought of it as whatever the user had typed in.  We might also keep other identifiers though as provided by Simbad.

So `name` would be editable and we do need unique names within a target list or else editing will be ambiguous.  That means the API will have to generate errors if you try to add two "NGC 5949" science targets for example.

File: docs/unshareTargets.md [hunk 4]
```diff
@@ -0,0 +1,396 @@
...
+targets by ID, we'll instead use target names and require unique names within a target list.
+
+(Since the `name` field will be needed to identify a target to edit, this raises the question of 
+how to change the target name.  Unfortunately it seems like a special `newName` edit field will be 
+required.)
+
+
+## Asterisms and Targets
+   
+I propose to remove the asterism as a separate type.  Observations instead have a list of
```
[the remark this answers is not in the document]
Comment (swalker2m) [reply docs/unshareTargets.md:20]: All observations would have a science target list.  Not all will have additional asterism details like those you'll see in GHOST observations.

File: docs/unshareTargets.md [hunk 5]
```diff
@@ -0,0 +1,396 @@
...
+
+
+## Asterisms and Targets
+   
+I propose to remove the asterism as a separate type.  Observations instead have a list of 
+science targets, accessible via a `science` field in the `targets` object.
+
+The `Targets` object would house everything related to targets, including the science target(s), 
+guide stars, blind offset target, etc.  The `science` field in `Targets` may return no targets, 
+one target, or multiple targets. We could add an optional `firstScience` field as well.  It
```
[the remark this answers is not in the document]
Comment (swalker2m) [reply docs/unshareTargets.md:25]: I was thinking that the base position concept apart from the science target(s) could apply to any observation.  Usually it wouldn't be specified and would be calculated as the center of the science target group.  If you wanted to use a different slew coordinate though, I don't see why we shouldn't allow it.

**Comment by toddburnside** [comment]
I will add that I'm still not sure what we're gaining by not sharing targets. I understand the benefits for constraints: when creating a new observation you can assign a sensible default constraint set and it simplified the API. For targets, there doesn't seem to be a sensible default. Is the API getting simpler?

**Comment by cquiroz** [comment]
I think the main issue we want to solve by not sharing is to avoid the situation where you edit properties in one observation and accidentally edit another.

**Comment by swalker2m** [comment]
> I will add that I'm still not sure what we're gaining by not sharing targets.

It's a fair question, and one that I found myself asking as well.  Consistency is the main thing, I think.  It seemed to me that it will be easier for script writers if you do the same kind of thing to edit (bulk or otherwise) constraints, targets, and configurations.

**Comment by swalker2m** [comment]
Since it's just a doc, I'll merge this PR.
